File: game.py
import random

def get_n_random_unique_numbers(n, max_value):
    return random.sample(range(max_value + 1), n)

def get_random_from_array(arr):
    return random.choice(arr)

def get_random_from_array_with_weights(values, weights):
    return random.choices(values, weights, k=1)[0]

class Game:
    def __init__(self):
        self.NUMBER_OF_CELLS = 16
        self.NUMBER_OF_ROWS_AND_COLUMNS = int(self.NUMBER_OF_CELLS ** 0.5)

        self.NEW_TILES_VALUES = [2, 4]
        self.NEW_TILES_WEIGHTS = [0.75, 0.25]

        self.tile_values = []
        self.score = 0


        self.init_grid()
        self.init_tiles()
        self.update_score()

    def init_grid(self):
        for _ in range(self.NUMBER_OF_CELLS):
            self.tile_values.append(0)

    def init_tiles(self):
        choosen_cells = get_n_random_unique_numbers(2, self.NUMBER_OF_CELLS - 1)
        for cell_number in choosen_cells:
            self.tile_values[cell_number] = get_random_from_array_with_weights(
                self.NEW_TILES_VALUES,
                self.NEW_TILES_WEIGHTS
            )

    def obter_celulas_vazias(self):
        return [index for index, num in enumerate(self.tile_values) if num == 0]

    def generate_new_tile(self):
        new_cell = get_random_from_array(self.obter_celulas_vazias())
        tile_value = get_random_from_array_with_weights(
            self.NEW_TILES_VALUES,
            self.NEW_TILES_WEIGHTS
        )
        self.tile_values[new_cell] = tile_value

    def update_score(self):
        self.score = sum(self.tile_values)

    def find_leftmost_available_tile(self, pos, value):
        tiles_left = pos % self.NUMBER_OF_ROWS_AND_COLUMNS

        for _ in range(tiles_left):
            checking_pos = pos - 1
            if self.tile_values[checking_pos] != 0:
                if self.tile_values[checking_pos] == value:
                    return checking_pos
                return pos
            pos = checking_pos
        return pos

    def find_topmost_available_tile(self, pos, value):
        tiles_up = pos // self.NUMBER_OF_ROWS_AND_COLUMNS

        for _ in range(tiles_up):
            checking_pos = pos - self.NUMBER_OF_ROWS_AND_COLUMNS
            if self.tile_values[checking_pos] != 0:
                if self.tile_values[checking_pos] == value:
                    return checking_pos
                return pos
            pos = checking_pos
        return pos

    def find_rightmost_available_tile(self, pos, value):
        tiles_right = ((pos // self.NUMBER_OF_ROWS_AND_COLUMNS + 1) * 4 - pos - 1)

        for _ in range(tiles_right):
            checking_pos = pos + 1
            if self.tile_values[checking_pos] != 0:
                if self.tile_values[checking_pos] == value:
                    return checking_pos
                return pos
            pos = checking_pos
        return pos

    def find_bottommost_available_tile(self, pos, value):
        tiles_down = (self.NUMBER_OF_CELLS - pos) // self.NUMBER_OF_ROWS_AND_COLUMNS

        for _ in range(tiles_down):
            checking_pos = pos + self.NUMBER_OF_ROWS_AND_COLUMNS

            if checking_pos >= self.NUMBER_OF_CELLS:
                return pos

            if self.tile_values[checking_pos] != 0:
                if self.tile_values[checking_pos] == value:
                    return checking_pos
                return pos
            pos = checking_pos
        return pos

    def move_tile_to(self, start, destination):
        if start == destination:
            return

        value = self.tile_values[start]
        destination_value = self.tile_values[destination]
        new_value = value

        if destination_value == value:
            new_value = value * 2

        self.tile_values[start] = 0
        self.tile_values[destination] = new_value

    def move_left(self):
        number_of_moves = 0
        for pos in range(self.NUMBER_OF_CELLS):
            if self.tile_values[pos] == 0:
                continue
            destination = self.find_leftmost_available_tile(pos, self.tile_values[pos])
            if destination != pos:
                self.move_tile_to(pos, destination)
                number_of_moves += 1

        if number_of_moves > 0:
            self.generate_new_tile()

    def move_up(self):
        number_of_moves = 0
        for pos in range(self.NUMBER_OF_CELLS):
            if self.tile_values[pos] == 0:
                continue
            destination = self.find_topmost_available_tile(pos, self.tile_values[pos])
            if destination != pos:
                self. move_tile_to(pos, destination)
                number_of_moves += 1

        if number_of_moves > 0:
            self.generate_new_tile()

    def move_right(self):
        number_of_moves = 0
        for pos in range(self.NUMBER_OF_CELLS - 1, -1, -1):
            if self.tile_values[pos] == 0:
                continue
            destination = self.find_rightmost_available_tile(pos, self.tile_values[pos])
            if destination != pos:
                self.move_tile_to(pos, destination)
                number_of_moves += 1

        if number_of_moves > 0:
            self.generate_new_tile()

    def move_down(self):
        number_of_moves = 0
        for pos in range(self.NUMBER_OF_CELLS - 1, -1, -1):
            if self.tile_values[pos] == 0:
                continue
            destination = self.find_bottommost_available_tile(pos, self.tile_values[pos])
            if destination != pos:
                self.move_tile_to(pos, destination)
                number_of_moves += 1

        if number_of_moves > 0:
            self.generate_new_tile()

    def show_table(self):
        print(f"Pontuação: {self.score}")
        for column in range(self.NUMBER_OF_ROWS_AND_COLUMNS):
            for row in range(self.NUMBER_OF_ROWS_AND_COLUMNS):
                print(self.tile_values[(column * 4) + (row % 4)], end=" ")
            print("")

File: visualization/src/game.js
import { getRandomFromArray, getRandomFromArrayWithWeights } from "./random";

class Game {
    constructor(tileValues = null) {
        this.NUMBER_OF_CELLS = 16;
        this.NUMBER_OF_ROWS_AND_COLUMNS = Math.sqrt(this.NUMBER_OF_CELLS);
        
        this.NEW_TILES_VALUES = [2, 4];
        this.NEW_TILES_WEIGHTS = [0.75, 0.25];
        
        this.tileValues = tileValues ? [...tileValues] : [];
        this.score = 0;

        if (!tileValues) {
            this.initGrid();
            this.initTiles();
        }

        this.updateScore();
    }

    clone() {
        return new Game(this.tileValues);
    }

    setState(state) {
        this.tileValues = state;
        this.updateScore();
    }

    initGrid() {
        this.tileValues = Array(this.NUMBER_OF_CELLS).fill(0);
    }

    initTiles() {
        let emptyCells = [...Array(this.NUMBER_OF_CELLS).keys()];
        let chosenCells = emptyCells.sort(() => 0.5 - Math.random()).slice(0, 2);
        
        chosenCells.forEach(cellNumber => {
            this.tileValues[cellNumber] = getRandomFromArrayWithWeights(this.NEW_TILES_VALUES, this.NEW_TILES_WEIGHTS);
        });
    }

    getEmptyCells() {
        return this.tileValues.map((val, index) => val === 0 ? index : -1).filter(index => index !== -1);
    }

    generateNewTile() {
        let newCell = getRandomFromArray(this.getEmptyCells());
        let tileValue = getRandomFromArrayWithWeights(this.NEW_TILES_VALUES, this.NEW_TILES_WEIGHTS);
        this.tileValues[newCell] = tileValue;
        this.updateScore();
    }

    getPossibleActions() {
        let actions = new Set();
        
        for (let pos = 0; pos < this.NUMBER_OF_CELLS; pos++) {
            if (this.tileValues[pos] === 0) continue;
            if (this.findRightmostAvailableTile(pos, this.tileValues[pos]) !== pos) actions.add("right");
            if (this.findLeftmostAvailableTile(pos, this.tileValues[pos]) !== pos) actions.add("left");
            if (this.findTopmostAvailableTile(pos, this.tileValues[pos]) !== pos) actions.add("up");
            if (this.findBottommostAvailableTile(pos, this.tileValues[pos]) !== pos) actions.add("down");
        }
        
        return Array.from(actions);
    }

    getScore(){
        this.updateScore();
        return this.score;
    }

    updateScore() {
        this.score = this.tileValues.reduce((sum, val) => sum + val, 0);
    }

    findLeftmostAvailableTile(pos, value) {
        while (pos % this.NUMBER_OF_ROWS_AND_COLUMNS !== 0) {
            let newPos = pos - 1;
            if (this.tileValues[newPos] !== 0 && this.tileValues[newPos] !== value) break;
            pos = newPos;
        }
        return pos;
    }

    findTopmostAvailableTile(pos, value) {
        while (pos >= this.NUMBER_OF_ROWS_AND_COLUMNS) {
            let newPos = pos - this.NUMBER_OF_ROWS_AND_COLUMNS;
            if (this.tileValues[newPos] !== 0 && this.tileValues[newPos] !== value) break;
            pos = newPos;
        }
        return pos;
    }

    findRightmostAvailableTile(pos, value) {
        while ((pos + 1) % this.NUMBER_OF_ROWS_AND_COLUMNS !== 0) {
            let newPos = pos + 1;
            if (this.tileValues[newPos] !== 0 && this.tileValues[newPos] !== value) break;
            pos = newPos;
        }
        return pos;
    }

    findBottommostAvailableTile(pos, value) {
        while (pos < this.NUMBER_OF_CELLS - this.NUMBER_OF_ROWS_AND_COLUMNS) {
            let newPos = pos + this.NUMBER_OF_ROWS_AND_COLUMNS;
            if (this.tileValues[newPos] !== 0 && this.tileValues[newPos] !== value) break;
            pos = newPos;
        }
        return pos;
    }

    moveTileTo(start, destination) {
        if (start === destination) return;
        let value = this.tileValues[start];
        let destinationValue = this.tileValues[destination];
        let newValue = destinationValue === value ? value * 2 : value;
        this.tileValues[start] = 0;
        this.tileValues[destination] = newValue;
    }

    moveLeft() {
        for (let pos = 0; pos < this.NUMBER_OF_CELLS; pos++) {
            if (this.tileValues[pos] === 0) continue;
            let destination = this.findLeftmostAvailableTile(pos, this.tileValues[pos]);
            this.moveTileTo(pos, destination);
        }
    }

    moveUp() {
        for (let pos = 0; pos < this.NUMBER_OF_CELLS; pos++) {
            if (this.tileValues[pos] === 0) continue;
            let destination = this.findTopmostAvailableTile(pos, this.tileValues[pos]);
            this.moveTileTo(pos, destination);
        }
    }

    moveRight() {
        for (let pos = this.NUMBER_OF_CELLS - 1; pos >= 0; pos--) {
            if (this.tileValues[pos] === 0) continue;
            let destination = this.findRightmostAvailableTile(pos, this.tileValues[pos]);
            this.moveTileTo(pos, destination);
        }
    }

    moveDown() {
        for (let pos = this.NUMBER_OF_CELLS - 1; pos >= 0; pos--) {
            if (this.tileValues[pos] === 0) continue;
            let destination = this.findBottommostAvailableTile(pos, this.tileValues[pos]);
            this.moveTileTo(pos, destination);
        }
    }

    applyAction(action){
        switch (action){
            case "up":
                this.moveUp();
                this.generateNewTile();
                break;
            case "down":
                this.moveDown();
                this.generateNewTile();
                break;
            case "right":
                this.moveRight();
                this.generateNewTile();
                break;
            case "left":
                this.moveLeft();
                this.generateNewTile();
                break;
        }
    }

    showTable() {
        console.log(`Pontuação: ${this.score}`);
        for (let i = 0; i < this.NUMBER_OF_ROWS_AND_COLUMNS; i++) {
            console.log(this.tileValues.slice(i * 4, (i + 1) * 4).join(" "));
        }
    }
}

export default Game